### PR TITLE
KNOX-3026 - End-users can exclude certain services or roles from CM service discovery

### PR DESCRIPTION
**Pull request by smolnar82**
## What changes were proposed in this pull request?

I added two new gateway-level configurations:
- `gateway.cloudera.manager.service.discovery.excluded.service.types`
- `gateway.cloudera.manager.service.discovery.excluded.role.types`

Both configurations can store a collection of comma-separated Cloudera Manager service or role types that should be excluded from CM service discovery. They both default to an empty set, meaning full CM discovery.
During a CM service discovery attempt, those configurations are checked and based on the supplied values the discovery logic skips the given service(s) or role(s) discovery.

## How was this patch tested?

I added and updated existing JUnit tests.
Tested on a live CM cluster:

Screenshot 2024-03-28 at 13 14 01

After restarting Knox, I exported discovery-related logs into [cm_discovery.txt](https://github.com/apache/knox/files/14789431/cm_discovery.txt)
```
grep "discovery.cm" /var/log/knox/gateway/gateway.log > cm_discovery.txt
```
Here is the list of relevant log entries for this PR:
```
$ grep -i skipping cm_discovery.txt 
2024-03-28 12:15:25,967 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$2(352)) - Skipping service discovery: CORE_SETTINGS-1 (CORE_SETTINGS)
2024-03-28 12:15:25,968 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$2(352)) - Skipping service discovery: KUDU-1 (KUDU)
2024-03-28 12:15:25,968 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$2(352)) - Skipping service discovery: YARN-1 (YARN)
2024-03-28 12:15:25,968 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$2(352)) - Skipping service discovery: IMPALA-1 (IMPALA)
2024-03-28 12:15:26,313 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: HDFS-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:27,117 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: HBASE-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:28,326 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: HIVE-1-GATEWAY-6b773e2d140b3c8597330125d17829ce (GATEWAY)
2024-03-28 12:15:28,326 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: HIVE-1-GATEWAY-962722fbc5dd79be35fd4c71f50d9db4 (GATEWAY)
2024-03-28 12:15:28,326 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: HIVE-1-GATEWAY-70be993d2303e6d7523ef15597ea5a1c (GATEWAY)
2024-03-28 12:15:30,067 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-2 (GATEWAY)
2024-03-28 12:15:30,068 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-3 (GATEWAY)
2024-03-28 12:15:30,068 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:30,170 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SPARK_ON_YARN-GATEWAY-3 (GATEWAY)
2024-03-28 12:15:30,170 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SPARK_ON_YARN-GATEWAY-2 (GATEWAY)
2024-03-28 12:15:30,170 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SPARK_ON_YARN-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:30,538 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: KAFKA-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:31,943 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: HIVE_ON_TEZ-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:31,943 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: HIVE_ON_TEZ-GATEWAY-2 (GATEWAY)
2024-03-28 12:15:31,943 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: HIVE_ON_TEZ-GATEWAY-3 (GATEWAY)
2024-03-28 12:15:32,550 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: LIVY-GATEWAY-3 (GATEWAY)
2024-03-28 12:15:32,551 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: LIVY-GATEWAY-2 (GATEWAY)
2024-03-28 12:15:32,551 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: LIVY-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:32,974 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: OZONE-1-GATEWAY-70be993d2303e6d7523ef15597ea5a1c (GATEWAY)
2024-03-28 12:15:32,974 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: OZONE-1-GATEWAY-962722fbc5dd79be35fd4c71f50d9db4 (GATEWAY)
2024-03-28 12:15:32,974 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: OZONE-1-GATEWAY-6b773e2d140b3c8597330125d17829ce (GATEWAY)
2024-03-28 12:15:34,801 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-962722fbc5dd79be35fd4c71f50d9db4 (GATEWAY)
2024-03-28 12:15:34,801 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-6b773e2d140b3c8597330125d17829ce (GATEWAY)
2024-03-28 12:15:34,801 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-70be993d2303e6d7523ef15597ea5a1c (GATEWAY)
2024-03-28 12:15:38,121 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-2 (GATEWAY)
2024-03-28 12:15:38,121 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-3 (GATEWAY)
2024-03-28 12:15:38,121 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:38,187 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-962722fbc5dd79be35fd4c71f50d9db4 (GATEWAY)
2024-03-28 12:15:38,187 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-6b773e2d140b3c8597330125d17829ce (GATEWAY)
2024-03-28 12:15:38,187 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-70be993d2303e6d7523ef15597ea5a1c (GATEWAY)
2024-03-28 12:15:40,426 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-2 (GATEWAY)
2024-03-28 12:15:40,426 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-3 (GATEWAY)
2024-03-28 12:15:40,426 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:40,490 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-962722fbc5dd79be35fd4c71f50d9db4 (GATEWAY)
2024-03-28 12:15:40,490 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-6b773e2d140b3c8597330125d17829ce (GATEWAY)
2024-03-28 12:15:40,490 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-70be993d2303e6d7523ef15597ea5a1c (GATEWAY)
2024-03-28 12:15:46,313 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-2 (GATEWAY)
2024-03-28 12:15:46,313 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-3 (GATEWAY)
2024-03-28 12:15:46,313 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: SQOOPCLIENT-GATEWAY-1 (GATEWAY)
2024-03-28 12:15:46,377 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-962722fbc5dd79be35fd4c71f50d9db4 (GATEWAY)
2024-03-28 12:15:46,377 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-6b773e2d140b3c8597330125d17829ce (GATEWAY)
2024-03-28 12:15:46,377 INFO  discovery.cm (ClouderaManagerServiceDiscovery.java:lambda$4(431)) - Skipping role discovery: TEZ-1-GATEWAY-70be993d2303e6d7523ef15597ea5a1c (GATEWAY)
```

### CODE REVIEW

File: gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerServiceDiscovery.java [hunk 1]
```diff
@@ -356,6 +361,14 @@ private List<ApiService> getClusterServices(ServiceDiscoveryConfig serviceDiscov
         final ApiServiceList serviceList = servicesResourceApi.readServices(serviceDiscoveryConfig.getCluster(), VIEW_SUMMARY);
         services = serviceList == null ? new ArrayList<>() : serviceList.getItems();
 
+        services = services.stream().filter(service -> {
+          if (excludedServiceTypes.contains(service.getType())) {
```
Comment (pzampino) [inline gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerServiceDiscovery.java:365]: I'm wondering if we can avoid case-sensitivity issues with this. For example, if the case of the service type is mistyped in the config or changes on the CM side. Perhaps, we can lowercase the values prior to the comparison?

File: gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerServiceDiscovery.java [hunk 2]
```diff
@@ -422,6 +437,21 @@ private ApiRoleList getRoles(ServiceDiscoveryConfig serviceDiscoveryConfig, Role
     return roles;
   }
 
+  private ApiRoleList excludeRoles(ApiRoleList roles) {
+    if (roles == null || roles.getItems() == null) {
+      return roles;
+    }
+    final ApiRoleList filteredRoles = new ApiRoleList();
+    roles.getItems().forEach(role -> {
+      if (excludedRoleTypes.contains(role.getType())) {
```
Comment (pzampino) [inline gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerServiceDiscovery.java:446]: Same concern as I have with the services wrt case sensitivty.

File: gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerServiceDiscovery.java (hunk 1 above)
Comment (smolnar82) [reply gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerServiceDiscovery.java:365]: This is a good idea. Let me fix that and submit a new PR soon.

**Review by pzampino — APPROVED** [review]
LGTM. Thank you!

**Review by moresandeep — APPROVED** [review]
Looks good!

**Pull request merged**